rmf_schedule_visualizer/src/rmf_schedule_visualizer/ScheduleDataNode.cpp

```diff
@@ -0,0 +1,258 @@
+/*
+ * Copyright (C) 2021 Open Source Robotics Foundation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+*/
+
+#include <rmf_schedule_visualizer/ScheduleDataNode.hpp>
+
+#include <rmf_traffic_ros2/StandardNames.hpp>
+#include <rmf_traffic_ros2/Time.hpp>
+#include <rmf_traffic_ros2/Trajectory.hpp>
+#include <rmf_traffic_ros2/schedule/MirrorManager.hpp>
+
+#include <rmf_traffic_msgs/msg/negotiation_notice.hpp>
+#include <rmf_traffic_msgs/msg/negotiation_conclusion.hpp>
+
+#include <rclcpp/rclcpp.hpp>
+
+namespace rmf_schedule_visualizer {
+
+//==============================================================================
+class ScheduleDataNode::Implementation
+{
+public:
+
+  using ConflictNotice = rmf_traffic_msgs::msg::NegotiationNotice;
+  using ConflictConclusion = rmf_traffic_msgs::msg::NegotiationConclusion;
+
+  Implementation() {}
+  Implementation(const Implementation&) {}
+
+  struct Data
+  {
+    rmf_traffic_ros2::schedule::MirrorManager mirror;
+
+    Data(rmf_traffic_ros2::schedule::MirrorManager mirror_)
+    : mirror(std::move(mirror_))
+    {
+      // Do nothing
+    }
+  };
+
+  rclcpp::Subscription<ConflictNotice>::SharedPtr conflict_notice_sub;
+  rclcpp::Subscription<ConflictConclusion>::SharedPtr conflict_conclusion_sub;
+
+  std::unique_ptr<Data> data;
+  mutable std::mutex mutex;
+  std::unordered_map<
+    rmf_traffic::schedule::Version,
+    std::vector<rmf_traffic::schedule::ParticipantId>> conflicts;
+
+  std::shared_ptr<rmf_traffic_ros2::schedule::Negotiation> negotiation =
+    nullptr;
+
+  void start(Data data);
+};
+
+//==============================================================================
+std::shared_ptr<ScheduleDataNode> ScheduleDataNode::make(
+  const std::string& node_name,
+  rmf_traffic::Duration wait_time)
+{
+  using ConflictNotice = rmf_traffic_msgs::msg::NegotiationNotice;
+  using ConflictConclusion = rmf_traffic_msgs::msg::NegotiationConclusion;
+
+  const auto start_time = std::chrono::steady_clock::now();
+  std::shared_ptr<ScheduleDataNode> schedule_data(
+    new ScheduleDataNode(std::move(node_name)));
+
+  schedule_data->_pimpl->conflict_notice_sub =
```

Comment:
Consider moving some of the `Implementation` setup into that class to shorten this function.